src/modbus-tcp-server.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright © 2016 DEIF A/S
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#ifndef _MSC_VER
+# include <unistd.h>
+#endif
+#include <time.h>
+
+#if defined(_WIN32)
+# define OS_WIN32
+/* ws2_32.dll has getaddrinfo and freeaddrinfo on Windows XP and later.
+ * minwg32 headers check WINVER before allowing the use of these */
+# ifndef WINVER
+# define WINVER 0x0501
+# endif
+/* Already set in modbus-tcp.h but it seems order matters in VS2005 */
+# include <winsock2.h>
+# include <ws2tcpip.h>
+# define SHUT_RDWR 2
+# define TIME_UTC 1
+# define close closesocket
+#else
+# include <sys/select.h>
+# include <sys/socket.h>
+# include <netinet/in.h>
+# include <arpa/inet.h>
+# include <netdb.h>
+#endif
+
+#include "modbus.h"
+#include "modbus-private.h"
+#include "modbus-tcp-server.h"
+#include <pthread.h>
+#include <time.h>
+
+#include <modbus.h>
+
+#define MB_TCP_SRV_IDLE_TIMEOUT 60
+
+/* modbus client data */
+struct modbus_tcp_client_t {
+    int socket;
+    time_t last_update;
+    struct modbus_tcp_client_t* next;
+};
+
+/* mdbus server data */
+struct _modbus_tcp_server {
+    int server_socket;
+    fd_set refset;
+    fd_set rdset;
+    int16_t fdmax;
+    uint16_t max_connections;
+    uint16_t port;
+    uint32_t idle_time_sec;
+    uint32_t select_to_sec;
+    uint32_t select_to_usec;
+    uint8_t shutdown;
+
+    modbus_t* ctx;
+    /* Linked list of all current connections */
+    struct modbus_tcp_client_t* conn;
+};
+
+static void _modbus_tcp_srv_rm_cli(modbus_tcp_server_t* data,
+        struct modbus_tcp_client_t* link) {
+    struct modbus_tcp_client_t* sp = NULL;
+    struct modbus_tcp_client_t* tmp = NULL;
+
+    for (sp = data->conn; sp != NULL; sp = sp->next) {
+        if (sp == link) {
+            close(sp->socket);
+            FD_CLR(sp->socket, &data->refset);
+
+            /* If we are pointing to HEAD, remove head */
+            if (tmp == NULL) {
+                data->conn = sp->next;
+                free(sp);
+                break;
+            } else {
+                struct modbus_tcp_client_t* tmp_next = sp->next;
+
+                /* check if last link */
+                if (tmp_next == NULL) {
+                    tmp->next = NULL;
+                    free(sp);
+                } else {
+                    tmp->next = tmp_next;
+                    free(sp);
+                }
+            }
+        }
+        tmp = sp;
+    }
+}
+
+static void _modbus_tcp_srv_del_oldest_cli(modbus_tcp_server_t* data) {
+    struct modbus_tcp_client_t* cli_to_remove = NULL;
+    time_t oldest_time = 0x7FFFFFFF;
+
+    /* Remove client from list of connected clients*/
+    struct modbus_tcp_client_t* sp = NULL;
+    for (sp = data->conn; sp != NULL; sp = sp->next) {
```

Comment:
If data is NULL this will crash.

Reply:
This is checked in line 289, and function is static.